File: python/trustlayer/modules/fraud.py
"""FraudShield — server-side fraud evaluation."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Optional

from ..models import EvaluationResponse

if TYPE_CHECKING:
    from ..session import TrustSession


class FraudShield:
    """
    Evaluates fraud risk for a session.

    Usage::

        session = client.create_session(type="transaction")
        fraud = FraudShield(session)
        result = fraud.evaluate(transaction_amount=2500.0)
    """

    def __init__(self, session: "TrustSession") -> None:
        self._session = session

    def evaluate(
        self,
        *,
        transaction_amount: Optional[float] = None,
        transaction_currency: str = "USD",
        transaction_id: Optional[str] = None,
        user_id: Optional[str] = None,
    ) -> EvaluationResponse:
        """
        Evaluate fraud risk, optionally enriched with transaction context.
        """
        if transaction_amount is not None:
            self._session.track_event(
                "custom",
                {
                    "signal_type": "transaction",
                    "amount": transaction_amount,
                    "currency": transaction_currency,
                    "transaction_id": transaction_id,
                    "user_id": user_id,
                },
            )
        return self._session.evaluate(modules=["fraud"])

    def monitor_transaction(self, transaction_id: str) -> None:
        """Tag a transaction ID to this session for tracking."""
        self._session.track_event(
            "custom",
            {"signal_type": "transaction_monitor", "transaction_id": transaction_id},
        )

File: src/modules/fraud.ts
import { TrustSession } from "../session";
import { EvaluationResponse } from "../api/types";

export interface FraudEvaluationContext {
  transactionAmount?: number;
  transactionCurrency?: string;
  userId?: string;
  transactionId?: string;
}

export interface FraudResult {
  fraudProbability: number;
  riskScore: number;
  recommendation: "allow" | "review" | "block";
  factors: string[];
}

/**
 * FraudShield — detects fraud signals in a session.
 */
export class FraudShield {
  constructor(private readonly session: TrustSession) {}

  /**
   * Evaluates fraud risk for the current session, optionally enriched
   * with transaction context.
   */
  async evaluate(context?: FraudEvaluationContext): Promise<FraudResult> {
    if (context?.transactionAmount !== undefined) {
      await this.session.trackEvent("custom", {
        signal_type: "transaction",
        amount: context.transactionAmount,
        currency: context.transactionCurrency ?? "USD",
        transaction_id: context.transactionId,
        user_id: context.userId,
      });
    }

    const evaluation: EvaluationResponse = await this.session.evaluate();

    const risk = evaluation.risk_score.risk_score;
    const factors = evaluation.risk_score.factors;

    // Compute fraud probability from risk score
    const fraudProbability = Math.round((risk / 100) * 1000) / 1000;

    let recommendation: "allow" | "review" | "block";
    if (evaluation.recommendation === "block") {
      recommendation = "block";
    } else if (
      evaluation.recommendation === "verify" ||
      evaluation.recommendation === "review"
    ) {
      recommendation = "review";
    } else {
      recommendation = "allow";
    }

    return {
      fraudProbability,
      riskScore: risk,
      recommendation,
      factors,
    };
  }

  /**
   * Tags a transaction ID to the session for tracking.
   */
  async monitorTransaction(transactionId: string): Promise<void> {
    await this.session.trackEvent("custom", {
      signal_type: "transaction_monitor",
      transaction_id: transactionId,
      monitored_at: new Date().toISOString(),
    });
  }
}
